File: javascript/wooah_tec/210508_01.js
function solution(weights, head2head) {
  const length = weights.length;

  const winCnt = {};
  const heavyWeightWinCnt = {};
  for (let i = 0; i < length; i++) {
    winCnt[i] = 0;
    heavyWeightWinCnt[i] = 0;
  }

  const weightsdict = weights.reduce(
    (prev, value, i) => ({ ...prev, [i]: value }),
    {}
  );

  const weightsort = Object.entries(weightsdict).sort((a, b) => b[1] - a[1]);

  const weightidx = weightsort.map((value) => +value[0]);

  // console.log({ winCnt, weightsdict, weightsort, weightidx });
  for (let idx = 0; idx < head2head.length; idx++) {
    const vs = head2head[idx];
    for (let i = 0; i < length; i++) {
      if (vs[i] === "W") {
        winCnt[idx] += 1;
        if (weights[idx] < weights[i]) {
          heavyWeightWinCnt[idx] += 1;
        }
      }
    }
  }
  //   weightsdict {0: 60, 1: 70, 2: 60}
  // weightsort [(1, 70), (0, 60), (2, 60)]
  // weightidx [1, 0, 2]
  // weightidx [1, 0, 2]
  // heavyWeightdict {1: 0, 0: 0, 2: 0}
  // heavyWeightSort [(1, 0), (0, 0), (2, 0)]
  // heavyWeightidx [1, 0, 2]
  // winCntdict {1: 0, 0: 0, 2: 0}
  // winCntSort [(1, 0), (0, 0), (2, 0)]
  // winCntidx [1, 0, 2]

  const heavyWeightdict = {};
  for (const i in weightidx) {
    heavyWeightdict[i] = heavyWeightWinCnt[i];
  }

  const heavyWeightSort = Object.entries(heavyWeightdict).sort(
    (a, b) => b[1] - a[1]
  );
  const heavyWeightidx = heavyWeightSort.map((v) => v[0]);

  const winCntdict = {};
  const winCntDict2 = new Map();
  for (const value of heavyWeightidx) {
    // console.log({ value });
    winCntDict2[value] = winCnt[value];
    // winCntdict[value] = winCnt[value];
    winCntDict2.set(value, winCnt[value]);
  }

  console.log([...winCntDict2]);
  // const winCntSort = Object.entries(winCntdict).sort((a, b) => b[1] - a[1]);
  const winCntSort2 = [...winCntDict2].sort((a, b) => b[1] - a[1]);

  const winCntIdx = winCntSort2.map((v) => +v[0] + 1);

  console.log({
    heavyWeightWinCnt,
    heavyWeightdict,
    heavyWeightSort,
    heavyWeightidx,
    winCntdict,
    winCntDict2,
  });
  console.log({ winCntSort2 });
  console.log({
    //   heavyWeightSort,
    //   winCntdict,
    //   winCntSort,
    //   winCntIdx,
  });

  // weightsdict {0: 60, 1: 70, 2: 60}
  // weightsort [(1, 70), (0, 60), (2, 60)]
  // weightidx [1, 0, 2]
  // heavyWeightWinCnt defaultdict(<class 'int'>, {1: 0, 0: 0, 2: 0})
  // heavyWeightdict {1: 0, 0: 0, 2: 0}
  // heavyWeightSort [(1, 0), (0, 0), (2, 0)]
  // heavyWeightidx [1, 0, 2]
  // winCntdict {1: 0, 0: 0, 2: 0}
  // winCntSort [(1, 0), (0, 0), (2, 0)]
  // winCntidx [1, 0, 2]
  // [2, 1, 3]

  return winCntIdx;
}

// console.log(solution([50, 82, 75, 120], ["NLWL", "WNLL", "LWNW", "WWLN"]));
console.log(solution([145, 92, 86], ["NLW", "WNL", "LWN"]));

File: javascript/wooah_tec/210508_01.py
from collections import defaultdict


def solution(weights, head2head):
    length = len(weights)
    winCnt = defaultdict(int)

    heavyWeightWinCnt = defaultdict(int)
    weightsdict = {i: weight for i, weight in enumerate(weights)}

    print('weightsdict', weightsdict)
    weightsort = sorted(weightsdict.items(), key=lambda x: x[1], reverse=True)
    print('weightsort', weightsort)
    weightidx = [i[0] for i in weightsort]
    print('weightidx', weightidx)

    for idx, vs in enumerate(head2head):
        for i in range(length):
            if vs[i] == "W":  # 승률 계산
                winCnt[idx] += 1
                if weights[idx] < weights[i]:  # 자신보다 무거운사람 이긴 횟수
                    heavyWeightWinCnt[idx] += 1

    heavyWeightdict = {i: heavyWeightWinCnt[i] for i in weightidx}

    heavyWeightSort = sorted(heavyWeightdict.items(),
                             key=lambda x: x[1], reverse=True)
    # print('heavyWeightWinCnt', heavyWeightWinCnt)
    # print('weightidx', weightidx)
    # print('heavyWeightdict', heavyWeightdict)
    # print('heavyWeightSort', heavyWeightSort)

    heavyWeightidx = [i[0] for i in heavyWeightSort]
    # print('heavyWeightidx', heavyWeightidx)

    winCntdict = {i: winCnt[i] for i in heavyWeightidx}
    # print('winCntdict', winCntdict)

    winCntSort = sorted(winCntdict.items(), key=lambda x: x[1], reverse=True)
    # print('winCntSort', winCntSort)

    winCntidx = [i[0] for i in winCntSort]
    # print('winCntidx', winCntidx)

    return list(map(lambda x: x+1, winCntidx))


if __name__ == "__main__":
    # weights = [50, 82, 75, 120]
    # weights = [145, 92, 86]
    weights = [60, 70, 60]
    # head2head = ["NLWL", "WNLL", "LWNW", "WWLN"]
    # head2head = ["NLW", "WNL", "LWN"]
    head2head = ["NNN", "NNN", "NNN"]
    print(solution(weights, head2head))
